In [1]:
### Exploratory data analysis
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib as plt

In [2]:
d=pd.read_csv("https://raw.githubusercontent.com/Premalatha-success/Datasets/main/hotel_bookings.csv")

In [3]:
d.shape

(119390, 32)

In [4]:
d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [5]:
d.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

##### 488 missing values in country (dtype=object)

4 missing values in children (dtype=float)

16340 missing values in agent (dtype=float)

112593 missing values in company (dtype=float)

#### drop company column too many missing values, replacing missing values in country with mode
missing values in children and agent with median/mean

In [6]:
d=d.drop(["company"],axis=1)

In [7]:
d.shape

(119390, 31)

##### missing values

In [8]:
m1=d["children"].median()

In [9]:
m1

0.0

In [10]:
d['children'].replace(np.nan,m1,inplace=True)

In [11]:
d.isnull().sum()

hotel                                 0
is_canceled                           0
lead_time                             0
arrival_date_year                     0
arrival_date_month                    0
arrival_date_week_number              0
arrival_date_day_of_month             0
stays_in_weekend_nights               0
stays_in_week_nights                  0
adults                                0
children                              0
babies                                0
meal                                  0
country                             488
market_segment                        0
distribution_channel                  0
is_repeated_guest                     0
previous_cancellations                0
previous_bookings_not_canceled        0
reserved_room_type                    0
assigned_room_type                    0
booking_changes                       0
deposit_type                          0
agent                             16340
days_in_waiting_list                  0


In [12]:
m2=d['agent'].median()

In [13]:
m2

14.0

In [14]:
d['agent'].replace(np.nan,m2,inplace=True)

In [15]:
d.isnull().sum()

hotel                               0
is_canceled                         0
lead_time                           0
arrival_date_year                   0
arrival_date_month                  0
arrival_date_week_number            0
arrival_date_day_of_month           0
stays_in_weekend_nights             0
stays_in_week_nights                0
adults                              0
children                            0
babies                              0
meal                                0
country                           488
market_segment                      0
distribution_channel                0
is_repeated_guest                   0
previous_cancellations              0
previous_bookings_not_canceled      0
reserved_room_type                  0
assigned_room_type                  0
booking_changes                     0
deposit_type                        0
agent                               0
days_in_waiting_list                0
customer_type                       0
adr         

In [16]:
m1=d['country'].mode().values[0] ##### values[0] for first mode, as there can be multiple modes
m1

'PRT'

In [17]:
d['country'].replace(np.nan,m1,inplace=True)

In [18]:
d.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces 

In [19]:
##### Check for duplicate values
dup=d.duplicated()
print(dup.sum())

32020


In [20]:
d.drop_duplicates(inplace=True)

In [21]:
d.duplicated().sum()

0

In [22]:
d.shape

(87370, 31)

In [32]:
#### Outliers
d.boxplot(column=['lead_time'])

<AxesSubplot: >

In [33]:
def rem_out(col):
    sorted(col)
    q1,q3=col.quantile([0.25,0.75])
    iqr=q3-q1
    l_ran=q1-1.5*iqr
    u_ran=q1+1.5*iqr
    return l_ran,u_ran

In [25]:
u_lt,h_lt=rem_out(d['lead_time'])
d['lead_time']=np.where(d['lead_time']>h_lt,h_lt,d['lead_time'])
d['lead_time']=np.where(d['lead_time']<u_lt,u_lt,d['lead_time'])

In [26]:
d.boxplot(column=['lead_time'])

<AxesSubplot: >

In [34]:
d.boxplot(column=['previous_cancellations'])

<AxesSubplot: >

In [35]:
d.dtypes


hotel                              object
is_canceled                         int64
lead_time                         float64
arrival_date_year                   int64
arrival_date_month                 object
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                               object
country                            object
market_segment                     object
distribution_channel               object
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                 object
assigned_room_type                 object
booking_changes                     int64
deposit_type                       object
agent                             

In [47]:
columns=['hotel','arrival_date_month','reservation_status','reservation_status_date','customer_type','deposit_type','reserved_room_type','assigned_room_type','market_segment','distribution_channel','meal','country']
d=pd.concat([d,dummies],axis=1) ### 

In [49]:
dummies=pd.get_dummies(d[columns]) ### converting categorical data into dummies/numerical data

In [48]:
d

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0      182.0               2015   
1       Resort Hotel            0      182.0               2015   
2       Resort Hotel            0        7.0               2015   
3       Resort Hotel            0       13.0               2015   
4       Resort Hotel            0       14.0               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0       23.0               2017   
119386    City Hotel            0      102.0               2017   
119387    City Hotel            0       34.0               2017   
119388    City Hotel            0      109.0               2017   
119389    City Hotel            0      182.0               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  country_UMI  country_URY  \
0                          0       2  ...            0            0   
1                          0       2  ...            0            0   
2                          1       1  ...            0            0   
3                          1       1  ...            0            0   
4                          2       2  ...            0            0   
...                      ...     ...  ...          ...          ...   
119385                     5       2  ...            0            0   
119386                     5       3  ...            0            0   
119387                     5       2  ...            0            0   
119388                     5       2  ...            0            0   
119389                     7       2  ...            0            0   

       country_USA country_UZB country_VEN country_VGB  country_VNM  \
0                0           0           0           0            0   
1                0           0           0           0            0   
2                0           0           0           0            0   
3                0           0           0           0            0   
4                0           0           0           0            0   
...            ...         ...         ...         ...          ...   
119385           0           0           0           0            0   
119386           0           0           0           0            0   
119387           0           0           0           0            0   
119388           0           0           0           0            0   
119389           0           0           0           0            0   

        country_ZAF  country_ZMB country_ZWE  
0                 0            0           0  
1   

In [50]:
d.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0      182.0               2015               July   
1  Resort Hotel            0      182.0               2015               July   
2  Resort Hotel            0        7.0               2015               July   
3  Resort Hotel            0       13.0               2015               July   
4  Resort Hotel            0       14.0               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  country_UMI  \
0                        0                     0       2  ...            0   
1                        0                     0       2  ...            0   
2                        0                     1       1  ...            0   
3                        0                     1       1  ...            0   
4                        0                     2       2  ...            0   

   country_URY country_USA country_UZB country_VEN country_VGB  country_VNM  \
0            0           0           0           0           0            0   
1            0           0           0           0           0            0   
2            0           0           0           0           0            0   
3            0           0           0           0           0            0   
4            0           0           0           0           0            0   

   country_ZAF  country_ZMB country_ZWE  
0            0            0           0  
1            0            0           0  
2            0            0           0  
3            0            0           0  
4            0            0           0  

[5 rows x 1198 columns]

In [51]:
d.drop(columns,axis=1,inplace=True)

In [52]:
d

is_canceled  lead_time  arrival_date_year  arrival_date_week_number  \
0                 0      182.0               2015                        27   
1                 0      182.0               2015                        27   
2                 0        7.0               2015                        27   
3                 0       13.0               2015                        27   
4                 0       14.0               2015                        27   
...             ...        ...                ...                       ...   
119385            0       23.0               2017                        35   
119386            0      102.0               2017                        35   
119387            0       34.0               2017                        35   
119388            0      109.0               2017                        35   
119389            0      182.0               2017                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  children  babies  ...  country_UMI  \
0                          0       2       0.0       0  ...            0   
1                          0       2       0.0       0  ...            0   
2                          1       1       0.0       0  ...            0   
3                          1       1       0.0       0  ...            0   
4                          2       2       0.0       0  ...            0   
...                      ...     ...       ...     ...  ...          ...   
119385                     5       2       0.0       0  ...            0   
119386                     5       3       0.0       0  ...            0   
119387                     5       2       0.0       0  ...            0   
119388                     5       2       0.0       0  ...            0   
119389                     7       2       0.0       0  ...            0   

        country_URY  country_USA  country_UZB  country_VEN  country_VGB  \
0                 0            0            0            0            0   
1                 0            0            0            0            0   
2                 0            0            0            0            0   
3                 0            0            0            0            0   
4                 0            0            0            0            0   
...             ...          ...          ...          ...          ...   
119385            0            0            0            0            0   
119386            0            0            0            0            0   
119387            0            0            0            0            0   
119388            0            0            0            0            0   
119389            0            0            0            0            0   

        country_VNM  country_ZAF  country_ZMB  country_ZWE  
0                 0            0            0            0  
1                 0            0            0            0  
2                 0            0            0            0  
3                 0            0            0            0  
4                 0            0            0            0  
...             ...          ...          ...          ...  
119385            0            0            0            0  
119386      